File: cs50x/week6/sentimental-readability/readability.py
"""
Specifications:
1. Coleman-Liau index is computed as 0.0588 * L - 0.296 * S - 15.8
2. L = avg no of letters per 100 words and S = avg no of sentences per 100 words
3. should count no of letters, word and sentences
4. case insensitive
5. spaces means end of a word 
6. punctuation means end of a sentence
7. round the value from the algorithm
8. prints Grade {grade number}
9. when algorithm returns 16:w
or higher "Grade 16+" for less then 1 "Before Grade 1
"""

import re


def coleman_liau_index(text):
    # count letters, words, and sentences
    letters = sum(1 for char in text if char.isalpha())
    words = len(text.split())
    sentences = len(re.findall(r"[.!?]", text))

    # calculate L and S
    L = (letters / words) * 100
    S = (sentences / words) * 100

    # Apply the coleman-liau formula
    index = round(0.0588 * L - 0.296 * S - 15.8)

    # Determine the grade level to print
    if index >= 16:
        print("Grade 16+")
    elif index < 1:
        print("Before Grade 1")
    else:
        print(f"Grade {index}")


# prompt the user for text input
text = input("Enter text: ")
coleman_liau_index(text)

File: cs50x2024/week2/readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>
int main()
{
    // defining variables used to store no of words sentences and letters
    float letters = 0, sentences = 0,
          words = 1; // last word doesnt have a space after it so take 1
    // promp user for input
    string text = get_string("TEXT: ");

    // loop through each character in text
    for (int i = 0, length = strlen(text); i < length; i++)
    {
        if (isalpha(text[i]))
        {
            letters++;
        }
        else if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            sentences++;
        }
        else if (isspace(text[i]))
        {
            words++;
        }
    }

    // average words per 100 words
    float L = letters / words;
    L *= 100;
    float S = sentences / words;
    S *= 100;

    // calculate Coleman-Liau index
    int grade = round(0.0588 * L - 0.296 * S - 15.8);

    // print grade lvl
    if (grade < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (grade >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", grade);
    }
    return 0;
}